cmake/python_functions.cmake

```diff
@@ -0,0 +1,30 @@
+#############################################################################
+# Copyright (c) 2018 Balabit
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+
+find_program(PEP8_EXECUTABLE NAMES pep8 pycodestyle HINTS ${CMAKE_BINARY_DIR}/.venv/bin)
+find_program(NOSETESTS_EXECUTABLE nosetests HINTS ${CMAKE_BINARY_DIR}/.venv/bin)
+find_program(PYLINT_EXECUTABLE pylint HINTS ${CMAKE_BINARY_DIR}/.venv/bin)
```

Comment:
Can you explain how to use the virtualenv? (What commands should I execute on a clean build directory)

What I don't understand is: if I want to use the python-venv target, I need to specify what python version I want to use. So before `make python-venv`, I need to run `cmake [...] -DPYTHON_VERSION=3`. But if I execute cmake without venv existing beforehand, the pep8, nosetests, pylint detection will fail during that, because the directories in HINTS do not exist. So in the end, it will pick up the python2 version from the system. The problem should exist in the autotools version.

As I think, we cannot reliably create python-venv target, because it depends on the configure/cmake due to python version, but it also needs to modify behaviour of configure/cmake due to pep8 nosetests etc (some kind of chicken-egg problem).

Reply:
my understanding was that find_program() would try the hints and then the normal executables without a path. Isn't this the behavior?

Reply:
Yes, but the problem is the time of the detection. As an example: in this case PEP8_EXECUTABLE will be filled when you call `cmake`. During that, indeed cmake will try the hint. But that time, we do not have `.venv`, so it will fall back to the global version. Then cmake will cache it, and it will not change even if you reexecute cmake.
From the [documentation](https://cmake.org/cmake/help/v3.4/command/find_path.html):
```
 If the file in a directory is found the result is stored in the variable and the search will not be repeated unless the variable is cleared
```

This is probably something different than in autotools. In case of autotools
```
PYTHON = $(or $(wildcard $(PYTHON_VENV_DIR)/bin/python),$(PYTHON))
```
autotools will generate a makefile with:
````
Makefile
4146:VPYTHON = $(or $(wildcard $(PYTHON_VENV_DIR)/bin/python),$(PYTHON))
```
so in the end the expression will probably recalculated each time you call make, thus causing no problem.